src/System.Windows.Forms.Primitives/src/System/Windows/Forms/Internals/UnsafeNativeMethods.cs

```
@@ -22,38 +19,29 @@ internal static class UnsafeNativeMethods
-#pragma warning restore CA1838 // Avoid 'StringBuilder' parameters for P/Invokes
+        public static extern unsafe int GetModuleFileName(HandleRef hModule, char* buffer, int length);
 
-        public static StringBuilder GetModuleFileNameLongPath(HandleRef hModule)
+        public static unsafe string GetModuleFileNameLongPath(HandleRef hModule)
         {
-            StringBuilder buffer = new StringBuilder(Kernel32.MAX_PATH);
-            int noOfTimes = 1;
-            int length = 0;
-            // Iterating by allocating chunk of memory each time we find the length is not sufficient.
-            // Performance should not be an issue for current MAX_PATH length due to this change.
-            while (((length = GetModuleFileName(hModule, buffer, buffer.Capacity)) == buffer.Capacity)
-                && Marshal.GetLastWin32Error() == ERROR.INSUFFICIENT_BUFFER
-                && buffer.Capacity < Kernel32.MAX_UNICODESTRING_LEN)
+            Span<char> buf = stackalloc char[Kernel32.MAX_UNICODESTRING_LEN];
+            fixed (char* valueChars = buf)
             {
-                noOfTimes += 2; // Increasing buffer size by 520 in each iteration.
-                int capacity = noOfTimes * Kernel32.MAX_PATH < Kernel32.MAX_UNICODESTRING_LEN ? noOfTimes * Kernel32.MAX_PATH : Kernel32.MAX_UNICODESTRING_LEN;
-                buffer.EnsureCapacity(capacity);
+                // Allocating a bigger buffer each time the function fails by MAX_PATH after starting
+                // with MAX_PATH is not possible with stack allocations, as such we must allocate exactly
+                // Kernel32.MAX_UNICODESTRING_LEN of memory so that way we never fail in the first place.
+                _ = GetModuleFileName(hModule, valueChars, buf.Length);
+                if (Marshal.GetLastWin32Error() == ERROR.INSUFFICIENT_BUFFER)
+                {
+                    // failure getting the module file names. Throw???
```

Given the assumption that we expect no further `INSUFFICIENT_BUFFER` error, we should throw error and fail fast or return appropriately.

---

What would you think the error message should be in this case? ``"Error getting module file names."`` perhaps?